*: Added PoI v2

That3Percent opened the pull request:
# PoI v2

The new hash function comes with some bells and whistles...

## Cryptographically strong
This is a cryptographically strong PoI suitable for on-chain rewards claims and bisecting.

Cyptographically strong in this context means:
- No collisions. Given any schema (or any schema evolution makes backward-compatible changes) it should not be possible to find a collision for a valid value for those schema(s). This means that the bisect algorithm will always work. This also means that it's not possible to attest to some reward claim and also to a different database simultaneously.
- No information leak. Given a hash, a schema, and some strict subset of fields for which values are known, it should be as hard for an attacker to derive the unknown values as brute-forcing the state space for those values. Furthermore, no subset of the unknown values can be derived. This is a weaker version of the following point...
- No derived hashes. Given a hash, a schema, and some strict subset of fields for which values are known, an attacker cannot swap out a known value for another value and produce a valid hash of the whole set of values. This means that even though an attacker knows the indexer id being mixed in for an on-chain claim, they cannot re-move it and re-add their own indexer id and produce a correct hash of the data - even if they have the hash for the version with and without the indexer id.

## Self-documenting
There is both an online and offline version of the PoI hash. The offline version exists in `proof_of_indexing/reference.rs` and serves as documentation for the hash, as well as a testament to the correctness of the online implementation. The question now of "what is this hash of" is simple to answer with the reference version - it's just the stable-hash of this struct containing all of the blocks, with each block containing all of the events for that block.

## Massively parallel
For the most part (hand-wavy details about unordered sets omitted) every member of every vec, every single field, and all data which contributes to the hash can be added to the hash out of order - so long as you know where it belongs in the final product. This opens the door for adding parallelism in the future. This also means that beyond the normal facilities for backward compatible additions that stable-hash provides, it may be possible to do make not backwards compatible changes with a database migration that does not involve re-indexing because we can update hashes in-place mixing in new data in a non-linear fashion.

## Fixed some stable hash impls
There was a pattern in some some enums to hash the discriminant using the provided sequence number, instead of a child sequence number. Indeed, this is a pattern that several implementations used (vec for example finishes with the count using the provided sequence number after calling next_child() n times). It turns out this is dangerous, because calls to next_child() modify the sequence number - meaning possibly fields couldn't be added to enum variants since this would mean in effect a different sequence number is passed to the discriminant breaking backwards compatibility. It also potentially complicates the online implementation since adding fields in one path (or members of a vec for instance) changes the sequence number in subtle ways that are not observed in every case) It turns out by luck that in none of the cases that were fixed the problem manifested in any meaningful way, but I didn't figure that out until after. Now these impls should be better examples of how to do this. Documentation is probably warranted on best practices for `impl StableHash`, but I'm behind schedule.

## Fix for PoI beyond indexed block
PoI includes block hash, and returns None if the requested block is beyond the latest indexed block. Also, everything is in a transaction.

# Other minor features / changes
- `Bytes` derefs as `[u8]`
- Moved some methods out of HostExports which did not make use of `&self`. For example, `big_int_to_string` does not need a host.
- Added `MaybeOwnedMut` this goes in a separate repository `maybe-owned` which may be reviewed and merged eventually

Jannis commented on store/postgres/src/store.rs:
```diff
@@ -997,40 +999,86 @@ impl StoreTrait for Store {
-                    Ok((causality_region, poi))
-                })
-                .collect::<Result<HashMap<_, _>, Error>>()?;
+                cancel.check_cancel()?;
```
We could already leave the transaction scope here, to release the lock on the table earlier, since we no longer need access to the database. The gain would be insignificant now, but with many causality regions, putting the final PoI together could take some time.

That3Percent replied:
Finishing the hash is practically instant on the database time scale, but I'll fix this anyway.

That3Percent commented on graph/src/components/subgraph/proof_of_indexing/event.rs:
```diff
@@ -0,0 +1,79 @@
+                id,
+                data,
+            } => {
+                entity_type.stable_hash(sequence_number.next_child(), state);
+                id.stable_hash(sequence_number.next_child(), state);
+                data.stable_hash(sequence_number.next_child(), state);
```
Even though this was accounted for, your comment shows you are asking the right kind of questions.

That3Percent commented on graph/src/components/subgraph/proof_of_indexing/online.rs:
```diff
@@ -0,0 +1,207 @@
+
+impl BlockEventStream {
+    fn new(block_number: u64) -> Self {
+        let events = traverse_seq_no(&[
+            1,                                // kvp -> v
+            0,                                // CausalityRegion.blocks: Vec<Block>
+            block_number.try_into().unwrap(), // Vec<Block> -> [i]
+            0,                                // Block.events -> Vec<ProofOfIndexingEvent>
+        ]);
```
I'll step back a bit to discuss sequence numbers generally first.

Recall that `StableHash` does not write any data for fields that have a default value. It does not write the integer `0` for example. Without a `SequenceNumber`, writing data in this fashion creates trivial collisions in common cases.

A `SequenceNumber` uniquely identifies where some piece of data lives in the structure. In the code above, this expresses "The first member of the block_number'th member of the first member of the second member" (Read backward).  Members can be fields in a struct, elements of a `Vec`, etc.

The crypto stable hash is interesting in that it's just a huge multi-set hash. That means that any member can be added to the set in any order. Members of the set that are fed into this hash are all of the "SequenceNumber + Data" pairs.

So, this code and the code like it identifies where the data lives (using the reference implementation as canon) then calls `stable_hash` on the data with that sequence number.

In that way we build up a hash equivalent to the hash of the reference structure, but with one piece of data at a time.

Jannis replied:
I think I understand. I'd find the following comment helpful (me):
```rust
// Obtain the sequence number that corresponds to the events in this stream.
// This sequence number is relative to a `(name, region)` tuple. The sequence
// traverses to the region in the tuple, the blocks in the region, the block with the
// block number that this stream is for, and then the events member in in that block.
let events = traverse_seq_no(&[
    1,                                // (name, region).1 — the causality region itself
    0,                                // region.blocks — blocks of the causality region (0th member of the region struct)
    block_number.try_into().unwrap(), // region.blocks[block_number] — the block the stream corresponds to
    0,                                // region.blocks[block_number].events — the events in the block (0th member of the block struct)
]);
```

Once you understand the sequence numbers, this may be obvious and we don't have to add comments like this to all `traverse_seq_no` calls, that would be excessive.

Jannis replied:
Perhaps this could become a comment on `traverse_seq` that uses this particular example to illustrate how it works.

That3Percent replied:
I have pushed the comment with a complete example of a struct, data, and field to identify

That3Percent commented on graph/src/components/subgraph/proof_of_indexing/online.rs:
```diff
@@ -0,0 +1,207 @@
+        }
+    }
+
+    pub fn pause(mut self, prev: Option<&[u8]>) -> Vec<u8> {
```
Yes. This gives us the state of the hash function in a way that is resumable. There are 2 cases where the hash function is resumed - when asking for the PoI, and combining with the next block.

Jannis replied:
That could make for a nice comment above the function 😉

That3Percent replied:
Added.

That3Percent commented on graph/src/components/subgraph/proof_of_indexing/online.rs:
```diff
@@ -0,0 +1,207 @@
+            per_causality_region: HashMap::new(),
+        }
+    }
+    /// Adds an event to the digest of the ProofOfIndexingStream local to the DataSource
```
Fixed

That3Percent commented on graph/src/components/subgraph/proof_of_indexing/online.rs:
```diff
@@ -0,0 +1,207 @@
+            0, // CausalityRegion.blocks: Vec<Block>
+        ]);
+        // + 1 is for
+        (self.block_number + 1).stable_hash(blocks_seq_no, &mut state);
```
Great question. That is redundant. I had the `block_ptr` field be of type EthereumBlockPointer - so that's where one of the block numbers came from. The other (this one, with the `+ 1` comment) is to delineate the end of the `events` Vec on `Block<'a>`. I'll fix it to just change the `block_ptr` field to `block_hash` and fix the comment.

That3Percent commented:
@Jannis I believe I have addressed all of your comments. I also found that the indexer address was incorrectly specified as `H256`, instead of `Address`. Please update your review.

Jannis commented:
@That3Percent Great, thanks for the awesome comment about the sequence number traversal. 

It looks like Travis isn't happy with the tests yet though.

That3Percent commented:
> It looks like Travis isn't happy with the tests yet though.
@Jannis Doh!. Fixed.

leoyvens commented on runtime/wasm/src/module/mod.rs:
```diff
@@ -124,11 +127,11 @@ pub(crate) struct WasmiModule {
 }
 
-impl WasmiModule {
+impl<'a> WasmiModule<'a> {
```
For `wasmtime` keeping this static is going to be a requirement.

The reason is that the current `wasmtime` API requires host exports to implement [IntoFunc](https://docs.rs/wasmtime/0.16.0/wasmtime/trait.IntoFunc.html), which is implemented only for static closures. Host exports need this of course so so I'll need to pass it into the closure, probably as an `Rc<RefCell<WasmiModule>>`.

That3Percent replied:
I pushed a change to make this `'static`. Not super happy about the change, since it turns a compile-time error into a runtime error. There wasn't a clear alternative though. Please update your review.
